src/hypothesis/extra/datetime.py

```diff
@@ -15,111 +15,412 @@
+                        low = max(low, getattr(min_dt, name))
+                    if constrained_high:
+                        high = min([high, getattr(max_dt, name)])
+                    if name == 'year':
+                        val = cu.centered_integer_range(data, low, high, 2000)
+                    else:
+                        val = cu.integer_range(data, low, high)
+                    result[name] = val
+                    constrained_low = constrained_low and val == low
+                    constrained_high = constrained_high and val == high
+                return dt.datetime(tzinfo=min_dt.tzinfo, **result)
+            except ValueError:
+                # We tried to draw a non-existent time, e.g. Feb 31
+                continue
+
+    def do_draw(self, data):
+        if self.min_dt.tzinfo == self.max_dt.tzinfo:
+            # No need to pick a timezone; they're the same (or both naive)
+            min_dt = self.min_dt
+            max_dt = self.max_dt
+        else:
+            # Choose a timezone to work in on this draw
+            if cu.boolean(data):
+                min_dt = self.min_dt
+                try:
+                    max_dt = self.max_dt.astimezone(self.min_dt.tzinfo)
+                except OverflowError:  # pragma: no branch  # pytest internals
+                    max_dt = self.min_dt.tzinfo.localize(dt.datetime.max)
+            else:
+                max_dt = self.max_dt
+                try:
+                    min_dt = self.min_dt.astimezone(self.max_dt.tzinfo)
+                except OverflowError:
+                    min_dt = self.max_dt.tzinfo.localize(dt.datetime.min)
+
+        result = self._draw_naive(data, min_dt, max_dt)
+        if (not self.timezones) or (self.allow_naive and cu.boolean(data)):
+            assert self.min_dt.tzinfo is None
+            return result
+        timezones = list(self.timezones)
+        for _ in range(len(self.timezones)):
+            try:
+                to_tz = cu.choice(data, timezones)
+                if self.min_dt.tzinfo is None:
+                    return to_tz.normalize(to_tz.localize(result))
+                else:
+                    return to_tz.normalize(result.astimezone(to_tz))
+            except OverflowError:
+                timezones.remove(to_tz)
+        raise InvalidState('Reached impossible state')  # pragma: no cover
 
 
 @defines_strategy
-def datetimes(allow_naive=None, timezones=None, min_year=None, max_year=None):
+def datetimes(allow_naive=None, timezones=None, min_year=None, max_year=None,
+              min_datetime=None, max_datetime=None):
     """Return a strategy for generating datetimes.
 
     allow_naive=True will cause the values to sometimes be naive.
     timezones is the set of permissible timezones. If set to an empty
-    collection all timezones must be naive. If set to None all available
+    collection all timezones will be naive. If set to None all available
     timezones will be used.
```

Comment:
I realise this was inherited from the previous API – I still find this mess of arguments quite confusing. In particular, the fact that there seem to be at least three ways I might determine whether it’s timezone-aware or naive.

Long-term, I think this might be a simpler API:

* Keep the `[min|max]_datetime` arguments, but coerce them to UTC as soon as they come in.
* Drop the `allow_naive` argument.
* If the user supplies a list of `timezones`, those are the available choices for the values. Otherwise, all values are naive.

Not something to change in this patch, but thoughts?

(Also, how can a time<em>zone</em> be native?)